### DISCUSSION TITLE
On the road to PyScaffold 4.0

### DISCUSSION THREAD

**Discussion by FlorianWilhelm** [pyscaffold/pyscaffold#235] [General]
## General discussion

This issue is a stub for everything that needs to be done and general discussion for the preparation of a PyScaffold 4.0 release which will include breaking changes.

**Comment by FlorianWilhelm** [comment]
@abravalheri , let's discuss here what needs to be done to get the `v4.0.x` branch building again as well as the new extensions for django and cookiecutter. I think it makes sense to have `setup_requires` in a way that extensions either work with PyScaffold 3.2 or 4.0 to avoid version conflicts when we include breaking changes. Do you agree?

To fix the unit tests I already released an alpha version of PyScaffold 4.0 and I think I also need to do this for all the current extensions we have like dsproject, pyproject etc. I would do so by just choosing a higher version number and also releasing an alpha or beta version with corresponding `setup_requires` for PyScaffold 4.0. This would avoid the problems we are having right now with the PyScaffold 4.0 unit tests. Do you agree?

**Comment by abravalheri** [comment]
Hi @FlorianWilhelm, great thinking!

Do you mean putting a cap on setup_requires (like you are doing nowadays for 3.2)? Something like: `setup_requires = pyscaffold>=4.0a0,<4.1a0`. I think this should be fine... Should we do `<4.1a0` or `<5.0a0`? Since are following semantic versioning, all the extensions that work on 4.1 will automatically work until 5.0. If we start with the cap in 5.0, when we release the next minor versions of PyScaffold, we can keep bumping the minimal. This way, we don't need to keep going to the extensions and update them.

Regarding the roadmap, I think we can address #215 and #219 relatively easily, because we have already a plan for that.

#218 is a little more tedious, since we have to go through all the extensions making sure the `__init__.py` files are empty and creating new releases.

#138 would be great to have. I have a proposal on how to rewrite `--update` as a built-in extension, but since what I envision it is back compatible, we can just go ahead with the initial plan.

#191 is more complicated though... In order to be able to have an interactive front-end (or something like an UI), we have to dissociate the concept of `parameters` and the internal implementation with `argparser`, and this might break the documented APIs. I will elaborate more in the open issue.

**Comment by abravalheri** [comment]
As a new improvement, #236 would be excellent to have (however, since it is non-breaking improvement, it can be shipped in a minor version bump) 😄

**Comment by FlorianWilhelm** [comment]
@abravalheri you are right about the cap in `setup_requires`, i.e. `setup_requires = pyscaffold>=4.0a0,<4.1a0`. Maybe from version 4.0 on we should switch to `<5.0a0` by default and just increase the major version more often whenever we break the extension API (which we should only do for good reasons.). This would also mean that we need to change `get_install_requires_version` in the `custom_extension` extension to increase the major version. Do you think we should also have a pyscaffold 4.0 branch for every extension? Just to make sure we don't mix up things? I really think it's gonna take us a while until we have a proper version 4.0 and in the meantime we might need to have fixes in the extensions for PyScaffold 3.2.

**Comment by abravalheri** [comment]
Hi @FlorianWilhelm, I have included 2 proposals as PR's recently that involve changing the API #294 and #295 (the first one is not exactly backwards compatible, although it can be made *sort-of* compatible), and 2 more as issues #296 and #297.

If you agree I can try to work on them (although it would be nice to have the help of @smartsammler for the gitlab-ci stuff).

I am postponing the easy changes (e.g. #266) and the `pyproject.toml` bomb when I feel that most of the API changes are done.

#236 is basically functional on the v4.0.x branch (although incomplete: there is no ``--config`` and ``--save-config`` flag). But implementing it after 4.0 is launched should be incremental and not break any API.

**Comment by FlorianWilhelm** [comment]
Hi @abravalheri, sounds good to me. I am also fine with introducing breaking changes for version 4, so #294 is fine. Seems like a major overhaul of PyScaffold and I am looking forward to it. Sorry, that I am of not of much help but I got quite a lot on my plate right now. I just want to quote Antoine de Saint-Exupery who said 

> Perfection is achieved, not when there is nothing more to add, but when there is nothing left to take away.

This doesn't mean *primitive obsession* as you can have elegant data structures to have an elegant solution. One just should always be careful to not overcomplicate things and add every feature that someone asked for.

That being said, I totally trust you and am excited for what's to come in version 4.0 :-)

**Comment by abravalheri** [comment]
Thanks @FlorianWilhelm! So I will proceed and before the release we can organise a review/discussion (it will be easier to remove stuff once everything is in place and we have a general overview)

Regarding the changes in API, to be fair I have the impression it actually simplifies stuff... I we skip the additions in terms of documentation (which I went a bit overboard), I have the impression the implementation shrank a bit (I deleted 2 huge complicated functions in the updates module to add 3 small and simple ones. After that change was made, adding the feature requested was a matter of 4 lines of code that actually do something (which to be sincere doesn't even need to be there... but I found it to be a nice re-usable addition)

**Comment by abravalheri** [comment]
I think we are getting closer to v4, finally! Most of the issues assigned in the road map are closed, indeed there are just a few new ones that I opened while implementing and thinking about the code.

It seems to me that #309 is worthy handling before v4, so PyScaffold is slightly more future-proof, and there are my 2 proposals about tox and gitlab pending for consideration.
I might go forward with the tox proposal, since there will be a escape hatch (`--no-tox`) and we are just flipping the logic between opt-in to opt-out, and I feel like this would encourage good practices.

**Comment by FlorianWilhelm** [comment]
Hi @abravalheri, 

I think using tox really is a good practice and it's better to add it as default, therefore I find your `--no-tox` idea good. Since Gitlab-CI gets more and more important I think having it directly as an internal extension makes sense as you suggested. 

Also, 4.0 is a great opportunity to get rid of old stuff like pkg-resources #309, which means I totally agree with you that it might make sense to handle it before. In general, sticking with semantic versioning and only breaking things in major versions like 4.0 would be my suggestion but please decide as you see fit. 

It's also great to see that you simplify PyScaffold. I hope it will make maintainability easier in the future and maybe we find even more contributors by keeping the entrance hurdle low.

Thank you. Is there anything else I forgot?

**Comment by abravalheri** [comment]
That was pretty much it @FlorianWilhelm thanks for the comments, I will keep the issues updated (probably I am going to slow down a bit the next weeks, but I hope we can have v4 in a few months).

I am about to release an alpha for 3.3, so that will be also nice :)

**Comment by FlorianWilhelm** [comment]
Should first the final version 3.3 be released first, then maybe 2-3 weeks waiting and seeing if anything pops up while at the same time testing some version 4.0 betas, followed by the 4.0 release?

**Comment by abravalheri** [comment]
I agree with this approach, it makes ton of sense. I published a v3.3 alpha a few days ago, but not so sure if we will have feedback on that one... I suppose the only thing left to add is the docs about `include_package_data` as we are discussing in #278?

I will try to release a v4 alpha this week, so we can also start fixing the extensions regarding the breaking changes in API.

**Comment by abravalheri** [comment]
Also, after adding the `include_package_data` stuff to the docs, I suppose it makes sense to make v4.0.x the master branch and create a v3.3.x

**Comment by FlorianWilhelm** [comment]
Is an interactive feature as described in #191 something that would lead to another breaking change or something that could be included in version 4.1?

**Comment by abravalheri** [comment]
That is a tricky one :P 

The way I understand this problem (and how I would give it a go) is:

> How can we implement the interactive feature without having to repeat everything specified for the CLI (argparse) and in a transparent way so new extensions are automatically added to the interactive prompt?

The best, but at the same time impossible way of doing that would be inspecting the `ArgumentParser` object dynamically and preparing the question/options presented to the user based on the `help` text for the CLI/default values/`argparse.Actions`. Unfortunately all the ways of doing that involve accessing private vars in the object (they are prefixed with an `_`). So the solutions we have are equally painful:

1. Replace argparse and change the APIs in a very much non backward compatible way
2. Still use argparse, but not directly, instead just collect the options and arguments in a nice data structure that can be used later to build both the argparser and the interactive stuff. Also a breaking change.
3. Inspect the private vars of argparse anyway, they have being there for ages and don't seem that will change very soon... This do not require breaking changes.
4. Do double-bookkeeping and have yet another API for the interactive stuff. So I guess we can say it is a backward compatible change? But not a nice way of doing that.
5. Put a proxy object on top of argparse that record the parameters in a special data structure and fallback to 2. In my opinion keeping the proxy in sync with the argparse API would be slightly less problematic then option 3 (the probability Python core members change it is slightly smaller), although more elegant.

I guess the only way of knowing is by starting the implementation.

**Comment by FlorianWilhelm** [comment]
Okay, do you think there is any real good reason to stay with `argparse` or use the 4.0 release to switch to a library that supports all this directly? Maybe there is one, I'll do a little research in the next days.

**Comment by FlorianWilhelm** [comment]
@abravalheri, okay, having checked your former comments and having looked a bit at `click` again, I see why in the case of PyScaffold it makes sense to stay with `argparse` to have a cleaner separation between parsing the args and using the results. I wish there would even be the possibility to even separate the actual parsing from the validation as this could be also used later if someone would use PyScaffold as a library. 

Have you seen https://pypi.org/project/argparse-prompt/ yet? I will check it out, maybe this is something we could use or apply a similar technique for the interactive mode in PyScaffold.

**Comment by abravalheri** [comment]
Hi @FlorianWilhelm, thanks for the link! I just had a look the guys working in the argparse-prompt where really clever to abuse the `type` machinery to do the interactive magic 😛 

In some way, it works more or less like the option 5 I was mentioning before... I wonder if it is safe to assume that the only method of argparse that is relevant for the "interativeness" is `add_argument`...

My quick and dirty idea (in 3) was just to inspect `ArgumentParser._actions` and use something like [inquirer](https://github.com/magmax/python-inquirer) or [PyInquirer](https://github.com/CITGuru/PyInquirer/) to sintetize the questions from the user... Yeah, it is bad, it is private ... but `argparse` objects do not offer other way of introspection (unless we write a layer on top)...

```python
>>> from pyscaffold import cli
>>> from pprint import pprint
>>> import argparse
>>> parser = argparse.ArgumentParser()
>>> cli.add_default_args(parser)
>>> pprint(parser._actions)
[_HelpAction(option_strings=['-h', '--help'], dest='help', nargs=0, const=None, default='==SUPPRESS==', type=None, choices=None, help='show this help message and exit', metavar=None),
 _StoreAction(option_strings=[], dest='project_path', nargs=None, const=None, default=None, type=None, choices=None, help='path where to generate/update project', metavar='PROJECT_PATH'),
 _StoreAction(option_strings=['-n', '--name'], dest='name', nargs=None, const=None, default=None, type=None, choices=None, help='installable name (as in `pip install`/PyPI, default: basename of PROJECT_PATH)', metavar='NAME'),
 _StoreAction(option_strings=['-p', '--package'], dest='package', nargs=None, const=None, default=None, type=None, choices=None, help='package name (as in `import`, default: NAME)', metavar='PACKAGE_NAME'),
 _StoreAction(option_strings=['-d', '--description'], dest='description', nargs=None, const=None, default=None, type=None, choices=None, help='package description', metavar='TEXT'),
 _StoreAction(option_strings=['-l', '--license'], dest='license', nargs=None, const=None, default=None, type=<function best_fit_license at 0x7f5fa8106620>, choices=['MIT', 'AGPL-3.0-or-later', 'AGPL-3.0-only', 'Apache-2.0', 'Artistic-2.0', 'BSD-2-Clause', 'BSD-3-Clause', 'CC0-1.0', 'EPL-1.0', 'GPL-2.0-or-later', 'GPL-2.0-only', 'GPL-3.0-or-later', 'GPL-3.0-only', 'ISC', 'LGPL-2.0-or-later', 'LGPL-2.0-only', 'LGPL-3.0-or-later', 'LGPL-3.0-only', 'MPL-2.0', 'Unlicense', 'Proprietary'], help='package license like MIT, AGPL-3.0-or-later, AGPL-3.0-only, Apache-2.0, Artistic-2.0, BSD-2-Clause, BSD-3-Clause, CC0-1.0, EPL-1.0, GPL-2.0-or-later, GPL-2.0-only, GPL-3.0-or-later, GPL-3.0-only, ISC, LGPL-2.0-or-later, LGPL-2.0-only, LGPL-3.0-or-later, LGPL-3.0-only, MPL-2.0, Unlicense, Proprietary (default: MIT)', metavar='LICENSE'),
 _StoreAction(option_strings=['-u', '--url'], dest='url', nargs=None, const=None, default=None, type=None, choices=None, help='main website/reference URL for package', metavar='URL'),
 _StoreTrueAction(option_strings=['-f', '--force'], dest='force', nargs=0, const=True, default=False, type=None, choices=None, help='force overwriting an existing directory', metavar=None),
 _StoreTrueAction(option_strings=['-U', '--update'], dest='update', nargs=0, const=True, default=False, type=None, choices=None, help='update an existing project by replacing the most important files like setup.py etc. Use additionally --force to replace all scaffold files.', metavar=None),
 _VersionAction(option_strings=['-V', '--version'], dest='version', nargs=0, const=None, default='==SUPPRESS==', type=None, choices=None, help="show program's version number and exit", metavar=None),
 _StoreConstAction(option_strings=['-v', '--verbose'], dest='log_level', nargs=0, const=20, default=None, type=None, choices=None, help='show additional information about current actions', metavar=None),
 _StoreConstAction(option_strings=['-vv', '--very-verbose'], dest='log_level', nargs=0, const=10, default=None, type=None, choices=None, help='show all available information about current actions', metavar=None),
 _StoreTrueAction(option_strings=['-P', '--pretend'], dest='pretend', nargs=0, const=True, default=False, type=None, choices=None, help='do not create project, but displays the log of all operations as if it had been created.', metavar=None),
 _StoreConstAction(option_strings=['--list-actions'], dest='command', nargs=0, const=<function list_actions at 0x7f5fa80f1f28>, default=None, type=None, choices=None, help='do not create project, but show a list of planned actions', metavar=None)]
```

(I did some checks, these APIs, despite private, have been pretty much stable for some long time)

Then use the technique I explained [#191](https://github.com/pyscaffold/pyscaffold/issues/191#issuecomment-511398276) to make `command` run the object generated by the inquirer.

**Comment by FlorianWilhelm** [comment]
Yeah, I also find it pretty neat what they do. Since I am on vacation the
next two weeks, I'll have some time to play around a bit and would push it
on some extra branch. Maybe we can then discuss if it makes sense.
Since they use GPL3 and it's really not much code I would rather use a
similar but independent approach and but it directly into Pyscaffold. Fine
by you?

**Comment by abravalheri** [comment]
Perfect, @FlorianWilhelm, I will leave that to you then and proceed with porting the extensions to v4!
Let me know if you need help with something.

**Comment by abravalheri** [comment]
🥳  [v4.0a2](https://pypi.org/project/PyScaffold/4.0a2/) is now on PyPI (with hopefully most of the breaking changes in terms of API).

**Comment by FlorianWilhelm** [comment]
@abravalheri I did some tests with `argparse-prompt` and although their trick is neat is doesn't work for many actions of `argparse`. For instance, it fails if you use `action='store_true'` as in this case there is no `type` argument allowed. Python 3.8 supports custom Actions and a similar trick as with `type` would be possible but we want to support Python >=3.6 and it doesn't feel clean anyway. 

I will now check out your solution 3 as it seems the best solution in that case. Reading a private/protected variable is fine I guess if one first inherits from the `ArgumentParser`.

**Comment by abravalheri** [comment]
I believe I ported all the extensions that needed to be ported to use the most updated version of PyScaffold's API. I also tagged  a bunch of them, but unfortunately I don't have access to all the projects in PyPI so I could not release alphas/betas for all of them. The situation is:

- [x] [pyscaffold](https://github.com/pyscaffold/pyscaffold) `v4.0a3`, development in the **v4.0.x** branch
- [x] [pyscaffoldext-django](https://github.com/pyscaffold/pyscaffoldext-django) `v0.1b3`, development in the **master** branch
- [x] [pyscaffoldext-cookiecutter](https://github.com/pyscaffold/pyscaffoldext-cookiecutter) `v0.1b3`, development in the **master** branch
- [x] [pyscaffoldext-markdown](https://github.com/pyscaffold/pyscaffoldext-markdown) `v0.4a3`, development in the **pyscaffold-v4** branch
- [x] [pyscaffoldext-custom-extension](https://github.com/pyscaffold/pyscaffoldext-custom-extension) `v0.6a2`, development in the **pyscaffold-v4** branch
- [x] [pyscaffoldext-dsproject](https://github.com/pyscaffold/pyscaffoldext-dsproject) `v0.5a1`, development in the **pyscaffold-v4** branch

(with the checkbox meaning uploaded to PyPI)

(sorry if I ended up changing the code too much, the new APIs in v4 allowed a bunch of simplifications ...)

**Comment by FlorianWilhelm** [comment]
Hi Anderson,

thanks so much for all your work. I'll change the permissions in the
evening as I am right now in the car on the way to a larger hike in
Austria.

When is was playing around a bit I managed to install Pyscaffold 4.0 with
the old extensions which lead to an error which was easily fixed by
uninstalling the extensions. That led me to the question if it would be
possible to have a somewhat nicer error message during the loading of an
extension that tells a user directly that there was a problem loading
extension XY.

Cheers, Florian

**Comment by abravalheri** [comment]
Hi @FlorianWilhelm, I tried to sketch something in #324, but I don't know how we could test that in the test suite though... Actually I am not even sure how to test that manually... Since we are coding the version requirements in the extensions' `install_requires`, isn't pip resolution algorithm supposed to prevent those problems to happen?

 I suppose the error happens  because old extensions try to import the `Extension` class from `pyscaffold.api` instead of `pyscaffld.extensions` right? Sorry if I changed the API... I decided to get rid of the `helpers` and the `utils` package and redistribute the functions and classes to places where they made more sense/were more meaningful.

I will leave the PR for you to check, please feel free to modify it. I will probably be away from the computer this weekend/next week, so you have some time to play around without me committing like crazy 😝 (I hope I am mostly finished though 🙏 , now all that should be left is fixing eventual bugs/adjustments).

**Comment by FlorianWilhelm** [comment]
> I believe I ported all the extensions that needed to be ported to use the most updated version of PyScaffold's API. I also tagged a bunch of them, but unfortunately I don't have access to all the projects in PyPI so I could not release alphas/betas for all of them. The situation is:
> 
> * [x]  [pyscaffold](https://github.com/pyscaffold/pyscaffold) `v4.0a3`, development in the **v4.0.x** branch
> * [x]  [pyscaffoldext-django](https://github.com/pyscaffold/pyscaffoldext-django) `v0.1b3`, development in the **master** branch
> * [x]  [pyscaffoldext-cookiecutter](https://github.com/pyscaffold/pyscaffoldext-cookiecutter) `v0.1b3`, development in the **master** branch
> * [x]  [pyscaffoldext-markdown](https://github.com/pyscaffold/pyscaffoldext-markdown) `v0.4a3`, development in the **pyscaffold-v4** branch
> * [x]  [pyscaffoldext-custom-extension](https://github.com/pyscaffold/pyscaffoldext-custom-extension) `v0.6a2`, development in the **pyscaffold-v4** branch
> * [x]  [pyscaffoldext-dsproject](https://github.com/pyscaffold/pyscaffoldext-dsproject) `v0.5a1`, development in the **pyscaffold-v4** branch
> 
> (with the checkbox meaning uploaded to PyPI)
> 
> (sorry if I ended up changing the code too much, the new APIs in v4 allowed a bunch of simplifications ...)

Thank you! I just added you as owner of all my pypi projects. I think it's better to be not the only owner :-)

**Comment by abravalheri** [comment]
Thanks @FlorianWilhelm and sorry for the delay, I just uploaded the missing alphas/betas.

**Comment by CarliJoy** [comment]
@abravalheri are the API Changes for v4.0 documented somewhere already?
I would be interested in updating my companies internal extension once v4.0 is released.

Thanks for all your work guys, btw :)

**Comment by abravalheri** [comment]
Hi @CarliJoy very good to know you are using PyScaffold :) I hope the experience have been a pleasant one.

I am trying to keep the documentation up to date as I change the code, so we have "hidden" docs for the new version on https://pyscaffold.org/en/v4.0.x/ (I am keeping it hidden because we are still working on the beta and therefore, it's subject to change).

As a quick example, you can check the new template for custom extensions:

https://github.com/pyscaffold/pyscaffoldext-custom-extension/blob/pyscaffold-v4/src/pyscaffoldext/custom_extension/templates/extension.template

I think the most important changes for extension authors are:
- the `helpers` modules is gone... now its functions are distributed in other modules with more "semantic" names (at least in my opinion :P)
- the dict struct representing the project  directory now does not include the project root folder
- `NO_OVERWRITE` and `NO_CREATE` flags were replaced with `pyscaffold.operations` function (roughly speaking `pyscaffold.operations import no_overwrite; NO_OVERWRITE = no_overwrite()`).

(You will find updated examples and descriptions on https://pyscaffold.org/en/v4.0.x/extensions.html)

I hope it helps

**Comment by CarliJoy** [comment]
@abravalheri thanks for your help.
Is there a timeline for the release of the 4.x version?
During Christmas I have some time off (you know lockdown and so one) and I could help out a bit bringing life to the 4.x version.
But I don't see any open issues regarding the 4.x version, so where would be a good place to start?
@FlorianWilhelm maybe you could help out here?

**Comment by abravalheri** [comment]
Hi @CarliJoy, thank you very much for your message.

I have been playing with version 4 lately and I consider it mostly stable (as always I am expecting some bugs to appear once we have more people using it for different workflows, but that is natural for a major version). I also released last month a beta version https://pypi.org/project/PyScaffold/4.0b1/.

In my mind we have a major issue and a minor issue left preventing the release:

1. We have an initial implementation for #191 (#325, #331), but that requires a change in API, that we are still not 100% secure. I even proposed an alternative "edit" extension which focus in solving the same problems as the "interactive" feature but in a different way (#333).

2. There is probably some errors in the CI configuration for [pyscaffoldext-django](https://travis-ci.org/github/pyscaffold/pyscaffoldext-django) and [pyscaffoldext-cookiecutter](https://travis-ci.org/github/pyscaffold/pyscaffoldext-cookiecutter), which means that the `travis` builtin extension needs some care. I don't think it is an error in PyScaffold code itself, but from time to time we always need to review the test environment configuration or fix problems in the test suites that never showed up before... Unfortunately I still haven't had the time to look into this one.

**Reply by CarliJoy** [reply]
I will look in the travis things.
About the #191 I will add my thought one it there.

**Reply by FlorianWilhelm** [reply]
Thanks @CarliJoy for looking into this!

**Comment by abravalheri** [comment]
After a long journey I think we are ready for a v4 release 🎉. I plan to release the last beta version soon, and after a couple of weeks the final version.

Between the beta and the v4.0 version, I would like to propose a feature freeze and just work in documentations, bugs and the extensions. At most, change the configuration of `setuptools_scm` to  `version_scheme = "no-guess-dev"` which matches better the way PyScaffold v3 used to behave (and is a [new option](https://github.com/pypa/setuptools_scm/#extending-setuptools_scm) in the latest versions of `setuptools_scm`).

Thank you everyone that participated on the efforts for the new version!

**Reply by abravalheri** [reply]
Version [v4.0b3](pypi.org/project/PyScaffold/4.0b3) released!

Now we wait a few weeks for bugs to appear before releasing v4 😝

**Comment by FlorianWilhelm** [comment]
After issue #378 is fixed, a new beta or even release candidate (`v4.0rc1`) for Pyscaffold and also the extension might be useful since there were a few minor changes lately and it would allow us to check the interplay of them again.

**Reply by abravalheri** [reply]
@FlorianWilhelm I was not planning to play with #376 for this release. If that is also true for you we can go for a release candidate. (My plans for PyScaffold before the final release are purely documentation: update README and maybe mention somewhere, like a *Why PyScaffold?* section, that the project is very mature and being used for more than 6 years in the wild).

**Reply by FlorianWilhelm** [reply]
Yes, I think one could move #376 to version 5.0 (or maybe 4.1) as it's not that important anyway for most people and I think it's more important to have the cool new features of 4.0 out in the wild soon. So I totally agree with what you say, only polishing docs and fixing bugs (if there are any). Good point with the *Why PyScaffold?* sections. If you release a release candidate of version 4.0, I would tweet about it and maybe even make a little LinkedIn Post inviting people to give it a test run.

**Reply by abravalheri** [reply]
Thank you @FlorianWilhelm, I am totally on board for that, I will release right now a beta, work on the *Why PyScaffold?* idea over the weekend, and then release the pre-release (nice to have the docs on a good shape when people come to see it from twitter/linkedin).

**Reply by FlorianWilhelm** [reply]
Thanks for working on this. Reading the `README.rst` again I wondered if we should mention the Cookiecutter template less prominently. I mean it's there and it seems to have even 2k downloads per month but I never heard of anyone using it nor any issues with it. Maybe one could stress more the point that with `pyscaffoldext-custom-extension` it's easy to write your own extension. People should not be nudged to cookiecutter for accomplishing this task.

**Reply by abravalheri** [reply]
Thanks for the tip @FlorianWilhelm, I hope to also have a look on that today and make a PR :P

**Comment by abravalheri** [comment]
🥳 [v4.0rc1](https://pypi.org/project/PyScaffold/4.0rc1/) is now on PyPI together with new releases for all the official extensions.

**Reply by abravalheri** [reply]
Thank you very much for the announcement on twitter @FlorianWilhelm. I also [published something to dev.to](https://dev.to/abravalheri/the-first-release-candidate-of-pyscaffold-4-0-just-arrived-5ecp).

Hopefully with the feedback and issues being reported we can have the final release soon.

**Comment by FlorianWilhelm** [comment]
@abravalheri, would you consider version 3.3 as still current after the release of 4.0 or as an old version? [with respect to the changelog]. I think many users might stick with version 3.x for a while, especially for projects created with it, thus showing that 3.3 is still maintained for a while might help them.

**Reply by abravalheri** [reply]
Hi @FlorianWilhelm thank you very much for bringing this up.
I think we should continue to support 3.3 for the time being, exactly for the reasons you mentioned, and also it is important to build a sense of trust in the community.

Let's keep it in the current versions section.

Should we release a 3.3.1 pinning maximum versions for the extensions in setup.cfg?

**Reply by FlorianWilhelm** [reply]
You mean pinning in the case someone uses `[all]` and so on? That could be done but I thought this is already asserted since the extensions themselves require PyScaffold>=4.0b2 in the current version? Nevertheless, it would be another security measure.

**Reply by abravalheri** [reply]
Exactly. It is just another security mechanism because at some point other beta versions of the extensions did not have that requirement.

I just released [v3.3.1](https://pypi.org/project/PyScaffold/3.3.1/) in PyPI. I also took the opportunity to add a little note in README and docs/index to clarify which versions of the documents the users should read (I noticed some people were working with v3.3 but reading v4 docs). @FlorianWilhelm, please feel free to change the notes.

**Comment by abravalheri** [comment]
Hello 👋 

It is been more than a month now that we had rc1 one and we have received very good feedback so far (or at least incorporated some changes in order to clarify users' doubts on v3.3) 🥳 .

Should we start to prepare for an official release? Maybe early March?

**Reply by FlorianWilhelm** [reply]
Hi @abravalheri 👋 

yes, I think PyScaffold v4 is ready to go and with two release candidates, we are already on the safe side. So whenever you want, please release it 😃  I am already super excited to see it finally released 🎉

**Reply by FlorianWilhelm** [reply]
@abravalheri, there is a bug in configupdater v1.1.3 when multi-line values contain empty lines and/or comments. I did fix the bug but had to rewrite quite a lot and thus I had to break the API, which leads to a version 2.0 soon. The good thing about this is that the code of configupdater became much cleaner and multi-line values containing comments and/or empty lines are now supported.

Right now I am waiting for some feedback in order to merge the PR. If this is done I will release version 2.0 soon, in about a week or so latest. I don't know if you want to wait with the release of PyScaffold 4.0 until it's tested with configupdater 2.0? I think it could also be done with version 4.0.1, so I have no strong opinion about it.

**Reply by abravalheri** [reply]
Hi @FlorianWilhelm, first of all, thank you very much for working in this fix, very good news.
I will have a look on your open PR, if the updates in ConfigUpdates does not force PyScaffold to any API changes I think we could go for the second approach to avoid postponing even more PyScaffold v4 (to be sincere I was about to release it this weekend but got side-tracked :P). I think we could do a v4.0 immediately followed by a v4.0.1b1, this way we still have some time for feedbacks on ConfigUpdater...

**Reply by abravalheri** [reply]
(In my mind there is nothing wrong with releasing v4 as a celebration of a cycle of work and then immediately start the next one :P)

**Reply by FlorianWilhelm** [reply]
Sounds great! So all that's left to do is to change the date in the changelog, commit and tag? Or have we forgotten to mention something in the changelog?

Looking forward to the release!

**Comment by abravalheri** [comment]
Thank you very much @FlorianWilhelm and everyone that contributed to this great release. 🥳 

I just tagged the master for v4, and it is already available on PyPI: https://pypi.org/project/PyScaffold/4.0/  🎉 

---

The following links are part of the release announcement:
- [dev.to](https://dev.to/abravalheri/pyscaffold-4-0-is-here-2p13)
- [medium](https://florian-wilhelm.medium.com/release-of-pyscaffold-v4-27d45f91905e)
- [twitter](https://twitter.com/PyScaffold/status/1367724839822368769)

@FlorianWilhelm also made a really cool video about the development of PyScaffold:

https://www.youtube.com/watch?v=0mhgIwJ-1ws

**Comment by jochengietzen** [comment]
Wow. The video looks amazing!

@FlorianWilhelm May I ask, how you generated this video? I've never seen such a visualisation for git!

**Reply by FlorianWilhelm** [reply]
The video was created with https://gource.io/ but it has problems with non-linearized repos and the visualization will be off in the sense that directories are shown which were already deleted. Thus:

1. Copy your repo to a tmp-repo as the next step will be destructive.
1.  linearize the repo to avoid Gource from showing files that no longer exists with:
     ```
     git filter-branch --parent-filter 'cut -f 2,3 -d " "'
     ```
1. Download avatars of the Maintainers from 

    - Gravatar: [https://gist.github.com/Blunderchips/35f5416fc4fe130efb68d1172448f774](https://gist.github.com/Blunderchips/35f5416fc4fe130efb68d1172448f774)
    - Github: [https://github.com/ty2k/github-avatar-downloader](https://github.com/ty2k/github-avatar-downloader)
   and specify them later with `--user-image-dir`.
   
1. Call Gource (but check the paths to the flags `--background-image`, `--user-image-dir`, `-i` before):
  ```
  gource -s .06 -1920x1080 --background-image gource/starry-sky-1654074_1920.jpg --date-format "Year %Y" --seconds-per-day 0.05 --user-image-dir gource/avatars --auto-skip-seconds .1 --multi-sampling --stop-at-end --highlight-users --hide mouse,progress,filenames --filename-time 600 --file-idle-time 0 --logo gource/logo.png --max-files 0 --background-colour 000000 --font-size 22 --output-ppm-stream - --output-framerate 60 | avconv -y -r 60 -f image2pipe -vcodec ppm -i - -b 65536K movie.mp4 && ffmpeg -i movie.mp4 -i gource/Chad_Crouch_-_Algorithms.mp3 -r 60 -c:a aac -vcodec libx264 -crf 24 output.mp4
  ```

**Reply by jochengietzen** [reply]
Thank you very much.

Appreciate the detailed explanation!
